File: packages/typescript/tests/glb-textures.test.ts
import { describe, it, expect } from 'vitest';
import * as fs from 'fs';
import * as path from 'path';
import { buildScene, toGLB } from '../src/index';
import { sniffImageMime } from '../src/model';

/**
 * Texture images in the GLB export.
 *
 * The parser has extracted texture bytes for a while, and every baked
 * primitive already carries UVs, but `toGLB` never wrote `images`/`textures`,
 * so an exported model lost its textures and fell back to each material's
 * averaged colour. Embedding them is opt-in: the images dominate the file
 * size, and a caller who only wants geometry should not pay for them.
 *
 * Fixture: capilla_quiroz_v17.skp, which carries three distinct JPEG textures
 * across four materials.
 */
function readFixture(name: string): ArrayBuffer {
  const buf = fs.readFileSync(path.join(__dirname, 'fixtures', name));
  return buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength) as ArrayBuffer;
}

/** Reads a GLB's JSON chunk (12-byte header + 8-byte chunk header). */
function glbJson(glb: Uint8Array): any {
  const view = new DataView(glb.buffer, glb.byteOffset, glb.byteLength);
  const jsonLen = view.getUint32(12, true);
  return JSON.parse(new TextDecoder().decode(glb.slice(20, 20 + jsonLen)));
}

/** Reads a GLB's BIN chunk. */
function glbBin(glb: Uint8Array): Uint8Array {
  const view = new DataView(glb.buffer, glb.byteOffset, glb.byteLength);
  const jsonLen = view.getUint32(12, true);
  const binStart = 20 + jsonLen;
  return glb.slice(binStart + 8, binStart + 8 + view.getUint32(binStart, true));
}

describe('sniffImageMime', () => {
  it('identifies PNG and JPEG by their magic bytes', () => {
    expect(sniffImageMime(new Uint8Array([0xff, 0xd8, 0xff, 0xe0, 1, 2]))).toBe('image/jpeg');
    expect(sniffImageMime(new Uint8Array([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]))).toBe('image/png');
  });

  it('rejects anything glTF cannot carry', () => {
    // e.g. TIFF, which older SketchUp files do contain
    expect(sniffImageMime(new Uint8Array([0x49, 0x49, 0x2a, 0x00, 1, 2, 3, 4]))).toBeNull();
    expect(sniffImageMime(new Uint8Array([]))).toBeNull();
  });
});

describe('buildScene texture collection', () => {
  const scene = buildScene(readFixture('capilla_quiroz_v17.skp'));

  it('collects the distinct texture images', () => {
    expect(scene.textures.length).toBe(3);
    for (const tex of scene.textures) {
      expect(tex.mimeType).toBe('image/jpeg');
      expect(tex.data.length).toBeGreaterThan(0);
      expect(typeof tex.filename).toBe('string');
    }
  });

  it('points textured materials at them', () => {
    const textured = (scene.gltfMaterials as any[]).filter(
      (m) => m.pbrMetallicRoughness?.baseColorTexture
    );
    expect(textured.length).toBe(4);
    for (const mat of textured) {
      const idx = mat.pbrMetallicRoughness.baseColorTexture.index;
      expect(idx).toBeGreaterThanOrEqual(0);
      expect(idx).toBeLessThan(scene.textures.length);
      // the resolved colour is kept alongside the image: glTF multiplies the
      // two, which is what SketchUp's colorized materials expect
      expect(mat.pbrMetallicRoughness.baseColorFactor).toHaveLength(4);
    }
  });

  it('keeps two textures apart when their average colours collide', () => {
    // Materials used to be deduplicated on colour alone, so two different
    // images that average to the same RGB collapsed into one material and one
    // of the images was lost. Real files do this - two fabrics both resolving
    // to 141,141,141 - so the texture has to be part of the key.
    const seen = new Set<number>();
    for (const mat of scene.gltfMaterials as any[]) {
      const tex = mat.pbrMetallicRoughness?.baseColorTexture;
      if (tex) seen.add(tex.index);
    }
    expect(seen.size).toBe(scene.textures.length);
  });

  it('does not fragment the mesh (same primitive count as before)', () => {
    // adding the texture to the grouping key must not split primitives that
    // were previously batched together
    expect(scene.glbPrimitives.length).toBe(21);
  });
});

describe('toGLB texture embedding', () => {
  const scene = buildScene(readFixture('capilla_quiroz_v17.skp'));

  it('omits the images by default', () => {
    const json = glbJson(toGLB(scene));
    expect(json.images).toBeUndefined();
    expect(json.textures).toBeUndefined();
    expect(json.samplers).toBeUndefined();
    // and leaves no dangling reference behind - a strict reader would reject it
    for (const mat of json.materials) {
      expect(mat.pbrMetallicRoughness.baseColorTexture).toBeUndefined();
    }
  });

  it('embeds them when asked', () => {
    const glb = toGLB(scene, { textures: true });
    const json = glbJson(glb);
    expect(json.images.length).toBe(3);
    expect(json.textures.length).toBe(3);
    expect(json.samplers.length).toBe(1);
    expect(json.materials.filter((m: any) => m.pbrMetallicRoughness?.baseColorTexture).length).toBe(4);

    // the bytes are really in the BIN chunk, not just declared
    const bin = glbBin(glb);
    for (let i = 0; i < json.images.length; i++) {
      const view = json.bufferViews[json.images[i].bufferView];
      const bytes = bin.slice(view.byteOffset, view.byteOffset + view.byteLength);
      expect(bytes.length).toBe(scene.textures[i].data.length);
      expect(bytes[0]).toBe(scene.textures[i].data[0]);
      // every bufferView must start 4-byte aligned
      expect(view.byteOffset % 4).toBe(0);
    }
  });

  it('produces a structurally valid GLB either way', () => {
    for (const glb of [toGLB(scene), toGLB(scene, { textures: true })]) {
      const view = new DataView(glb.buffer, glb.byteOffset, glb.byteLength);
      expect(view.getUint32(0, true)).toBe(0x46546c67); // "glTF"
      expect(view.getUint32(8, true)).toBe(glb.byteLength);
      const jsonLen = view.getUint32(12, true);
      expect(jsonLen % 4).toBe(0);
      const binStart = 20 + jsonLen;
      expect(view.getUint32(binStart + 4, true)).toBe(0x004e4942); // "BIN"
      const declared = glbJson(glb).buffers[0].byteLength;
      const chunkLen = view.getUint32(binStart, true);
      // the chunk may be padded up to 3 bytes past the declared length
      expect(declared).toBeLessThanOrEqual(chunkLen);
      expect(chunkLen - declared).toBeLessThan(4);
    }
  });

  it('is a no-op on a model with no textures', () => {
    const plain = buildScene(readFixture('Untitled.skp'));
    expect(plain.textures.length).toBe(0);
    expect(toGLB(plain, { textures: true })).toEqual(toGLB(plain));
  });
});

File: packages/cpp/tests/glb_textures_test.cpp
#include <algorithm>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <gtest/gtest.h>
#include <initializer_list>
#include <string>

#include <openskp/openskp.hpp>

#define TINYGLTF_NO_STB_IMAGE
#define TINYGLTF_NO_STB_IMAGE_WRITE
#include <tiny_gltf.h>

#include "test_helpers.hpp"

// GLB texture embedding and the material-identity fix it depends on
// (openskp#193, ported from TypeScript).
//
// Before this, gltf_materials was keyed on (color, double_sided) alone, so
// two different textures that happened to average to the same RGB would
// silently collapse into one material and lose an image. Fixed by keying on
// (color, double_sided, texture_index) instead, at both the face-grouping
// and material-dedup layers.
//
// Fixture: capilla_quiroz_v17.skp, which carries 3 real, distinct JPEG
// textures - real coverage, not a synthetic mock.

namespace openskp {
namespace {

// TinyGLTF is built with TINYGLTF_NO_STB_IMAGE, so it has no default pixel
// decoder and refuses to load a GLB with embedded images unless the caller
// supplies one. These tests only check structure (bufferView, mimeType,
// indices), never pixel data, so a no-op stand-in is enough.
bool skip_image_decode(tinygltf::Image*, const int, std::string*, std::string*, int, int,
                       const unsigned char*, int, void*) {
  return true;
}

tinygltf::Model load_glb(const ByteBuffer& bytes) {
  tinygltf::Model model;
  tinygltf::TinyGLTF loader;
  loader.SetImageLoader(skip_image_decode, nullptr);
  std::string error;
  std::string warning;
  EXPECT_TRUE(loader.LoadBinaryFromMemory(&model, &error, &warning, bytes.data(), bytes.size()))
      << error << warning;
  EXPECT_TRUE(error.empty()) << error;
  return model;
}

bool contains_bytes(const ByteBuffer& haystack, std::initializer_list<std::uint8_t> needle) {
  return std::search(haystack.begin(), haystack.end(), needle.begin(), needle.end()) !=
         haystack.end();
}

const ByteBuffer kJpegMagic{0xff, 0xd8, 0xff};

TEST(GlbTextures, SceneDeduplicatesTexturesAndKeysMaterialsByThem) {
  const auto scene = SkpFile::open(test::fixture("capilla_quiroz_v17.skp")).build_scene();

  ASSERT_EQ(scene.textures.size(), 3);
  for (const auto& tex : scene.textures) {
    EXPECT_TRUE(tex.mime_type == "image/jpeg" || tex.mime_type == "image/png");
    EXPECT_FALSE(tex.data.empty());
  }

  std::size_t textured = 0;
  for (const auto& mat : scene.gltf_materials) {
    if (mat.pbr_metallic_roughness.base_color_texture.has_value()) {
      ++textured;
      EXPECT_LT(*mat.pbr_metallic_roughness.base_color_texture, scene.textures.size());
    }
  }
  EXPECT_EQ(textured, 4);
}

TEST(GlbTextures, ExportOmitsImagesByDefault) {
  const auto scene = SkpFile::open(test::fixture("capilla_quiroz_v17.skp")).build_scene();
  const auto bytes = to_glb(scene);

  const std::string text(bytes.begin(), bytes.end());
  EXPECT_EQ(text.find("\"images\""), std::string::npos);
  EXPECT_FALSE(contains_bytes(bytes, {0xff, 0xd8, 0xff}));

  const auto model = load_glb(bytes);
  EXPECT_TRUE(model.images.empty());
}

TEST(GlbTextures, ExportEmbedsTexturesWhenAsked) {
  const auto scene = SkpFile::open(test::fixture("capilla_quiroz_v17.skp")).build_scene();
  const auto without_textures = to_glb(scene);
  const auto with_textures = to_glb(scene, GlbOptions{true});

  EXPECT_GT(with_textures.size(), without_textures.size());
  EXPECT_TRUE(contains_bytes(with_textures, {0xff, 0xd8, 0xff}));

  const auto model = load_glb(with_textures);
  ASSERT_EQ(model.images.size(), 3);
  for (const auto& image : model.images) {
    EXPECT_GE(image.bufferView, 0);
    ASSERT_GE(image.mimeType.size(), 6);
    EXPECT_EQ(image.mimeType.substr(0, 6), "image/");
  }
  ASSERT_EQ(model.textures.size(), 3);
  for (const auto& texture : model.textures) {
    EXPECT_GE(texture.source, 0);
  }

  bool found_textured_material = false;
  for (const auto& material : model.materials) {
    if (material.pbrMetallicRoughness.baseColorTexture.index >= 0) {
      found_textured_material = true;
      EXPECT_LT(material.pbrMetallicRoughness.baseColorTexture.index,
                static_cast<int>(model.textures.size()));
    }
  }
  EXPECT_TRUE(found_textured_material);
}

TEST(GlbTextures, ExportGlbFileWritesEmbeddedTextures) {
  const auto scene = SkpFile::open(test::fixture("capilla_quiroz_v17.skp")).build_scene();
  const auto output = std::filesystem::temp_directory_path() / "openskp-cpp-textures-test.glb";
  export_glb(scene, output, GlbOptions{true});

  std::ifstream stream(output, std::ios::binary | std::ios::ate);
  ASSERT_TRUE(stream);
  const auto size = stream.tellg();
  ByteBuffer actual(static_cast<std::size_t>(size));
  stream.seekg(0);
  stream.read(reinterpret_cast<char*>(actual.data()), size);
  stream.close();

  EXPECT_TRUE(contains_bytes(actual, {0xff, 0xd8, 0xff}));
  std::filesystem::remove(output);
}

}  // namespace
}  // namespace openskp
